src/app/consensus_test/consensus_test.ml

```diff
@@ -0,0 +1,229 @@
+    }
+  in
+  let candidate_consensus_state_with_hashes =
+    { With_hash.hash =
+        Mina_state.Protocol_state.hashes candidate_block.protocol_state
+    ; data =
+        Mina_state.Protocol_state.consensus_state candidate_block.protocol_state
+    }
+  in
+  match
+    Consensus.Hooks.select
+      ~context:(module Context)
+      ~existing:existing_consensus_state_with_hashes
+      ~candidate:candidate_consensus_state_with_hashes
+  with
+  | `Take ->
+      let candidate_length =
+        Mina_state.Protocol_state.consensus_state candidate_block.protocol_state
+        |> Consensus.Data.Consensus_state.blockchain_length
+        |> Unsigned.UInt32.to_int
+      in
+      let existing_length =
+        Mina_state.Protocol_state.consensus_state existing_block.protocol_state
+        |> Consensus.Data.Consensus_state.blockchain_length
+        |> Unsigned.UInt32.to_int
+      in
+      compare_lengths candidate_length existing_length
+  | `Keep ->
+      CandidateShorter
+
+let process_precomputed_blocks ~context ~current_chain blocks =
+  let%bind () =
+    Deferred.List.iter blocks ~f:(fun candidate_block ->
+        let existing_block = List.hd_exn !current_chain in
+        let select_outcome =
+          run_select ~context existing_block candidate_block
+        in
+        current_chain :=
+          update_chain ~current_chain ~candidate_block ~select_outcome ;
+        return () )
+  in
+  return ()
+
+let write_blocks_to_output_dir ~current_chain ~output_dir =
+  let sorted_output =
+    List.map ~f:precomputed_block_to_block_file_output !current_chain
+    |> List.rev
+  in
+  let write_block_to_file (i : int) (block : BlockFileOutput.t) :
+      unit Deferred.t =
+    let block_json_str =
+      block |> BlockFileOutput.to_yojson |> Yojson.Safe.to_string
+    in
+    let output_file = sprintf "%s/block_%d.json" output_dir i in
+    Writer.save output_file ~contents:block_json_str
+  in
+  let%bind () =
+    Deferred.List.iteri sorted_output ~f:(fun i block ->
+        let%bind () = write_block_to_file i block in
+        return () )
+  in
+  return ()
+
+let main () ~blocks_dir ~output_dir =
+  let logger = Logger.create () in
+  let current_chain : Mina_block.Precomputed.t list ref = ref [] in
+
+  [%log info] "Starting to read blocks dir"
+    ~metadata:[ ("blocks_dir", `String blocks_dir) ] ;
+  let%bind block_sorted_filenames = read_directory blocks_dir in
+  let%bind precomputed_blocks =
+    Deferred.List.map block_sorted_filenames ~f:(fun json ->
+        read_block_file json )
+  in
+  [%log info] "Finished reading blocks dir" ;
+  let context = context logger in
+
+  match precomputed_blocks with
+  | [] ->
+      failwith "No blocks found"
+  | first_block :: precomputed_blocks ->
+      [%log info] "Starting to process blocks"
+        ~metadata:[ ("num_blocks", `Int (List.length precomputed_blocks)) ] ;
+      current_chain := [ first_block ] ;
```

Comment:
NIT: since you always write this value before reading from the ref, it would read better to just put `let current_chain = ref [ first_block ]` on this line and not define the ref as initially empty above.